onnxruntime/core/optimizer/compute_optimizer/upstream_reshape_actors.cc

```diff
@@ -0,0 +1,287 @@
+
+  propagate_input_indices.clear();
+  all_input_cmp_rets.clear();
+  propagate_input_indices.reserve(current_node.InputDefs().size());
+  all_input_cmp_rets.reserve(current_node.InputDefs().size());
+
+  // Here we extend all input shapes to have the same length as the output shape (e.g. fully broadcasted shape).
+  // For each input, we check
+  //   1). besides the first two dims, all other dims are equal, and
+  //   2). the first two dims are compatible with the output shape.
+  //
+  // The dims are compatible as long as it meets one of the following conditions:
+  // 1. The first two dims do not exist (e.g. input shape is [1024] and output shape is [2, 16, 1024])
+  // 2. The first two dims exist, and are equal (e.g. input shape is [2, 16, 1024] and output shape is [2, 16, 1024])
+  // 3. [Not Supported] The first two dims exist, and the first dim is 1 (e.g. input shape is [1, 16, 1024] and output
+  //    shape is [2, 16, 1024])
+  // 4. [Not Supported] The first two dims exist, and the second dim is 1 (e.g. input shape is [2, 1, 1024] and output
+  //    shape is [2, 16, 1024])
+  // 5. [Not Supported] The first dim does not exist, and the second dim exist (can be 1 or not).
+  for (int input_idx = 0; input_idx < static_cast<int>(current_node.InputDefs().size()); ++input_idx) {
+    auto input_shape = current_node.InputDefs()[input_idx]->Shape();
+
+    auto [success, ret] = CompareInputShapeWithOutputShape(data_input_arg->Shape(),
+                                                           input_shape);
+
+    if (!success) {
+      LOG_DEBUG_INFO(logger, "Fail SimplePointwiseReshapeActor::PreCheck for node " + current_node.Name() +
+                                 ": reshape's data input rank < passthrough node's input rank");
+      return false;
+    }
+
+    if (ret.size() < 2) {
+      LOG_DEBUG_INFO(logger, "Fail SimplePointwiseReshapeActor::PreCheck for node " + current_node.Name() +
+                                 ": full broadcasted shape has dim < 2.");
+      return false;
+    }
+
+    all_input_cmp_rets[input_idx] = ret;
+
+    if (ret[0] == DimCompareRet::NotExist && ret[1] == DimCompareRet::NotExist) {
+      // Don't need to propagate Reshape since the two leading dims to flatten do not ext.
+      LOG_DEBUG_INFO(logger, "In SimplePointwiseReshapeActor::PreCheck for node " + current_node.Name() +
+                                 ": skip propagating for the input because the merged dims does not exist.");
+      continue;
+    } else if (ret[0] == DimCompareRet::Equal && ret[1] == DimCompareRet::Equal) {
+      propagate_input_indices.push_back(input_idx);
+    } else {
+      LOG_DEBUG_INFO(logger, "Fail SimplePointwiseReshapeActor::PreCheck for node " + current_node.Name() +
+                                 ": not supported cases for two leading dims check.");
+      return false;
+    }
+
+    // All other dims should be equal
+    for (int dim_index = 2; dim_index < static_cast<int>(ret.size()); ++dim_index) {
```

Comment:
with line:106, just need all dims to be equal?

Reply:
I did this by intention :) , in case we need later support more cases for the initial 2 dims (described in the comment above).

  // 3. [Not Supported] The first two dims exist, and the first dim is 1 (e.g. input shape is [1, 16, 1024] and output
  //    shape is [2, 16, 1024])
  // 4. [Not Supported] The first two dims exist, and the second dim is 1 (e.g. input shape is [2, 1, 1024] and output
  //    shape is [2, 16, 1024])
  // 5. [Not Supported] The first dim does not exist, and the second dim exist (can be 1 or not).